platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,754 @@
+
+    fn deref(&self) -> &Self::Target {
+        &*self.data
+    }
+}
+
+impl DerefMut for SmhcDataGuard<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.data
+    }
+}
+
+impl SmhcData {
+    /// Check for error bits in the interrupt status registers.
+    fn error_status(smhc: &smhc::RegisterBlock) -> Option<ErrorKind> {
+        let mint = smhc.smhc_mintsts.read();
+        let idst = smhc.smhc_idst.read();
+
+        if mint.m_dsto_vsd_int().bit_is_set() {
+            Some(ErrorKind::DataStarvationTimeout)
+        } else if mint.m_fu_fo_int().bit_is_set() {
+            Some(ErrorKind::FifoUnderrunOverflow)
+        } else if mint.m_cb_iw_int().bit_is_set() {
+            Some(ErrorKind::CommandBusyIllegalWrite)
+        } else if mint.m_rto_back_int().bit_is_set() {
+            Some(ErrorKind::ResponseTimeout)
+        } else if mint.m_rce_int().bit_is_set() {
+            Some(ErrorKind::ResponseCrc)
+        } else if mint.m_re_int().bit_is_set() {
+            Some(ErrorKind::Response)
+        } else if mint.m_dto_bds_int().bit_is_set() {
+            Some(ErrorKind::DataTimeout)
+        } else if mint.m_dce_int().bit_is_set() {
+            Some(ErrorKind::DataCrc)
+        } else if mint.m_dse_bc_int().bit_is_set() {
+            Some(ErrorKind::DataStart)
+        } else if mint.m_dee_int().bit_is_set() {
+            Some(ErrorKind::DataEnd)
+        } else if idst.des_unavl_int().bit_is_set() || idst.fatal_berr_int().bit_is_set() {
+            Some(ErrorKind::Dma)
+        } else {
+            None
+        }
+    }
+
+    fn advance_isr(&mut self, smhc: &smhc::RegisterBlock, num: u8) {
+        tracing::trace!(state = ?self.state, smhc = num, "SMHC{num} interrupt");
+
+        let mut needs_wake = false;
+
+        self.err = Self::error_status(smhc);
+        if let Some(err) = self.err {
+            // Clear all interrupt bits
+            smhc.smhc_rintsts.write(|w| unsafe { w.bits(0xFFFF_FFFF) });
+            smhc.smhc_idst.write(|w| unsafe { w.bits(0x3FF) });
+
+            needs_wake = true;
+        }
+
+        self.state = match self.state {
+            State::Idle => State::Idle,
+            State::WaitForCommand => {
+                if smhc.smhc_mintsts.read().m_cc_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.cc().set_bit());
+                    match self.op {
+                        SmhcOp::None => State::Idle,
+                        SmhcOp::Control => {
+                            needs_wake = true;
+                            State::Idle
+                        }
+                        SmhcOp::Read { .. } | SmhcOp::Write { .. } => State::WaitForDma,
+                    }
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDma => {
+                // TODO: better way to check for RX and TX,
+                // *normal interrupt summary* does not seem to work
```

Comment:
out of curiousity, what does this mean exactly?

Reply:
In the IDMAC interrupt status register has the `normal interrupt summary` as bit 8, which is supposed to be set when either of the RX or TX interrupt bits are set. That would be convenient to check if the operation we are waiting for has finished.
However, it doesn't work and the bit is never set.
Maybe the documentation of the interrupt enable register isn't complete in the manual version I have (v1.0).
If you have the newer version, could you let me know if `SMHC_IDIE_REG` has a bit 8 that can be set? :smile:

Reply:
i checked v1.4 of the D1 manual, and the table for `SMHC_IDIE_REG` lists the following bits:

| Bit | Name |
|:-----|:---------|
| 0 | `TX_INT_ENB` |
| 1 | `RX_INT_ENB` |
| 2 | `FERR_INT_ENB` |
| 4 | `DES_UNAVL_INT_ENB` |
| 5| `ERR_SUM_INT_ENB` |

Bit 3 and bits 6-31 are not listed in the table in that version of the manual. I assume this means that the "normal interrupt summary" bit is not actually real?

Reply:
Ah, it looks like v1.4 of the manual _does_ discuss a `NOR_INT_SUM` ("Normal Interrupt Summary") bit, but it's bit 8 of the `SMHC_IDST_REG` register, rather than `SMHC_IDIE_REG`.

The documentation for that bit says the following:
> NOR_INT_SUM (NIS)
>
> Normal Interrupt Summary
>
> Logical OR of the following:
> IDST[0]: Transmit Interrupt
> IDST[1]: Receive Interrupt
>
> Only unmasked bits affect this bit.
> This is a sticky bit and must be cleared each time a corresponding
> bit that causes NIS to be set is cleared.

Hope that's helpful!

Reply:
Ah okay, it's the same for the v1.0 manual. The IDST is the status register while the IDIE is the enable register.
Since the IDST talks about bit 8, but I don't see it getting set, I was wondering if it had to be enabled as well, and that the corresponding bit in the IDIE was missing in my manual version. But it doesn't seem so.

Reply:
Oh, I see. Hmm. The Allwinner docs are....not great... I wonder if they're trying to say that you need to be writing it back every time we read from the IDST reg to ensure it gets set on the next IRQ?